src/hooks/useGlobalSignIn.ts

```diff
@@ -0,0 +1,158 @@
+import { useEffect, useMemo, useState } from "react";
+import { v4 as uuidv4 } from "uuid";
+import every from "lodash/every";
+import isEmpty from "lodash/isEmpty";
+import {
+    adminGenericProxyFetch,
+    useDeskproAppClient,
+    useDeskproAppEvents,
+    useInitialisedDeskproAppClient
+} from "@deskpro/app-sdk";
+import { getQueryParams } from "../utils";
+import { getCurrentUserService } from "../services/teamviewer";
+import { Account, Settings, AuthTokens } from "../services/teamviewer/types";
+
+export const useGlobalSignIn = () => {
+    const { client } = useDeskproAppClient();
+    const [ settings, setSettings ] = useState<Settings|null>(null);
+    const [ callbackUrl, setCallbackUrl ] = useState<string|null>(null);
+    const [ poll, setPoll ] = useState<(() => Promise<{ token: string }>)|null>(null);
+    const [ isLoading, setIsLoading ] = useState<boolean>(false);
+    const [ isBlocking, setIsBlocking ] = useState<boolean>(true);
+    const [ accessCode, setAccessCode ] = useState<string|null>(null);
+    const [ user, setUser ] = useState<Account|null>(null);
+
+    const key = useMemo(() => uuidv4(), []);
+
+    const signOut = () => {
+        client?.setAdminSetting("");
+        setUser(null);
+        setAccessCode(null);
+    };
+
+    const signIn = () => {
+        poll && (async () => {
+            setIsLoading(true);
+            setAccessCode((await poll()).token)
+        })();
+    };
+
+    useDeskproAppEvents({
+        onAdminSettingsChange: setSettings,
+    }, []);
+
+    // Initialise OAuth flow
+    useInitialisedDeskproAppClient((client) => {
+        (async () => {
+            const { callbackUrl, poll } = await client.oauth2().getAdminGenericCallbackUrl(
+                key,
+                /code=(?<token>[\d\w]+)/,
+                /state=(?<key>.+)/
+            );
+
+            setCallbackUrl(callbackUrl);
+            setPoll(() => poll);
+        })();
+    }, [key]);
+
+    // Exchange auth code for auth/refresh tokens
+    useInitialisedDeskproAppClient((client) => {
+        const canRequestAccessToken = every([
+            accessCode,
+            callbackUrl,
+            settings?.client_id,
+            settings?.client_secret,
+        ]);
+
+        if (!canRequestAccessToken) {
+            return;
+        }
+
+        const url = new URL(`https://webapi.teamviewer.com/api/v1/oauth2/token`);
+
+        const requestOptions: RequestInit = {
+            method: "POST",
+            body: new URLSearchParams({
+                grant_type: "authorization_code",
+                code: accessCode as string,
+                client_id: settings?.client_id as string,
+                client_secret: settings?.client_secret as string,
```

Comment:
Never include secrets via the frontend, this must be passed as a proxy placeholder. `client_secret` must be a "backend only" setting in the manifest too.

Reply:
@AshleyDawson I copied this approach from the Salesforce app. I specified that it `isBackendOnly`. Will the placeholder work in `adminGenericProxyFetch`?

<img width="777" alt="Screenshot 2022-09-14 at 12 52 57" src="https://user-images.githubusercontent.com/11043120/190122995-bc009a98-aa25-48ae-b8af-b84cfde4c553.png">

Reply:
then it will have to be fixed in the Salesforce App too

Reply:
@zpawn sorry, this is my mistake - this is actually coming from the settings form itself and not from any persisted settings, please ignore